File: assay/src/core/format.ts
import type { Verdict } from './types';

/** Human/agent-readable one-block summary of a verdict. Host-agnostic. */
export function formatVerdict(v: Verdict): string {
  const score = v.acceptanceScore === null ? 'n/a' : `${Math.round(v.acceptanceScore * 100)}%`;
  const mark = (s: string) => (s === 'pass' ? '✓' : s === 'fail' ? '✗' : s === 'unresolved' ? '?' : '–');
  const lines = v.results.map(
    (r) => `  ${mark(r.status)} ${r.criterionId} [${r.status}]${r.status !== 'pass' ? ` — ${r.reason}` : ''}`,
  );
  const detail = v.applicable === 0
    ? '0 applicable — nothing was decided'
    : `${v.passed}/${v.applicable}; unresolved=${v.unresolved}`;
  const header = `assay · ${v.subject} · ${v.archetype} — ${v.outcome} · acceptance ${score} (${detail})`;
  return [header, ...lines].join('\n');
}

/**
 * The canonical machine form of a verdict: one NDJSON line, key order fixed by the
 * Verdict type. This is what CI artifacts and agent loops should persist/parse —
 * `formatVerdict` is the human face, this is the wire face.
 */
export function verdictToJsonLine(v: Verdict): string {
  return JSON.stringify(v);
}

File: assay.net/src/Assay.Net/Format.cs
using System.Text;

namespace Assay.Net;

/// <summary>Human/agent-readable rendering of verdicts — the .NET sibling of the TS formatVerdict.</summary>
public static class Format
{
    /// <summary>One-block summary: outcome + score, then one line per criterion.</summary>
    public static string Verdict(Verdict v)
    {
        var score = v.AcceptanceScore is null
            ? "n/a"
            : $"{(int)Math.Round(v.AcceptanceScore.Value * 100)}%";
        var sb = new StringBuilder();
        sb.Append("assay · ").Append(v.Subject).Append(" · ").Append(v.Archetype)
          .Append(" — ").Append(v.Outcome.ToString().ToLowerInvariant())
          .Append(" · acceptance ").Append(score)
          .AppendLine(v.Applicable == 0
              ? " (0 applicable — nothing was decided)"
              : $" ({v.Passed}/{v.Applicable}; unresolved={v.Unresolved})");
        foreach (var r in v.Results)
        {
            var mark = r.Status switch
            {
                VerdictStatus.Pass => "✓",
                VerdictStatus.Fail => "✗",
                VerdictStatus.Unresolved => "?",
                _ => "–",
            };
            sb.Append("  ").Append(mark).Append(' ').Append(r.CriterionId)
              .Append(" [").Append(StatusName(r.Status)).Append(']');
            if (r.Status != VerdictStatus.Pass) sb.Append(" — ").Append(r.Reason);
            sb.AppendLine();
        }
        return sb.ToString().TrimEnd();
    }

    private static string StatusName(VerdictStatus status) => status switch
    {
        VerdictStatus.NotApplicable => "not-applicable",
        _ => status.ToString().ToLowerInvariant(),
    };
}
